### PR TITLE
Fix for Issue with SumIf on "full columns" range #646

### LINKED ISSUE #646: Issue with SumIf on "full columns" range
## Read and complete the full issue template

**Do you want to request a *feature* or report a *bug*?**
- [x] Bug
- [ ] Feature

**Version of ClosedXML**

0.90

**What is the current behavior?**

Call cell.Value for a cell with SumIf with columns range parameters (e.g. =SUMIF(A:A;"=A";B:B)) doesn't work.
An exception ClosedXML.Excel.CalcEngine.ExpressionParseException is thrown when evaluating empty cells (issue 1 in code).
Another problem arise when cell is part of values range, in this case evaluation after a while hangs with OutOfMemoryException (Issue 2 in code).

**What is the expected behavior or new feature?**

The Value must be returned correctly.

**Did this work in previous versions of our tool?  Which versions?**

No

## Reproducibility

**Code to reproduce problem:**
```c#
public void Main()
{
    using (var wb = new XLWorkbook("TestSumIf.xlsx"))
            {
                // Issue 1 - ExpressionParseException thrown
                var ws = wb.Worksheet("SUMIF Not Same Columns");
                var value = ws.Cell("E1").Value; // ExpressionParseException 
                Console.WriteLine(value);
                // Issue 2 - Evaluation hangs
                //var ws = wb.Worksheet("SUMIF Same Columns");
                //var value = ws.Cell("B7").Value; // OutOfMemoryException 
                //Console.WriteLine(value);
            }
}
```
- [ ] I attached a sample spreadsheet.  (You can drag files on to this issue)
[TestSumIf.xlsx](https://github.com/ClosedXML/ClosedXML/files/1615541/TestSumIf.

## Fixing
I will prepare a pull request with issues fixed.

### PR DESCRIPTION
**Pull request by roby70**
Managed SumIf on string with spaces and applied evaluation of values only when condition is satisfied (resolve "recursion" problem due to use SumIf formula in same range for values).

#### What's this PR do?
Manage issues described in #646 fixing SumIf evaluation for full columns range.
I also change evaluation method for SumIf evaluating values to sum only when condition was satisfied.

#### Where should the reviewer start?
Changed CalcEngineHelpers.ValuesSatisfiesCriteria to manage empty string for values and string with spaces.
Changed MathTrig.SumIf to evaluate values to sum only when condition apply.
Inserted specific tests with sample workbook.

#### How should this be manually tested?
Same way as described in issue #646 on same sheet. 

Fixes #646

### CODE REVIEW

**Comment by igitur** [comment]
Also please resolve the conflict that has now emerged.

File: AppForIssueTracking/AppForIssueTracking.csproj [hunk 1]
```diff
@@ -0,0 +1,72 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
```
Comment (igitur) [inline AppForIssueTracking/AppForIssueTracking.csproj]: Please remove this project.

File: ClosedXML_Tests/Excel/CalcEngine/SumIfTests.cs [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+using System.Threading.Tasks;
+using ClosedXML.Excel;
+using NUnit.Framework;
+
+namespace ClosedXML_Tests.Excel.CalcEngine
+{
+    [TestFixture]
+    public class SumIfTests
+    {
+        [TestCase("SUMIF Not Same Columns", "E1", 3)]
```
[the remark this answers is not in the document]
Comment (roby70) [reply ClosedXML_Tests/Excel/CalcEngine/SumIfTests.cs]: I found SumIf definition in MathTrig class, so I suppose test class must be MathTrigTests and not StatisticalTests.
Is it correct?

[the remark this answers is not in the document]
Comment (igitur) [reply ClosedXML_Tests/Excel/CalcEngine/SumIfTests.cs]: Sorry, my bad. You're right. `MathTrigTests` is the correct class.

**Comment by roby70** [comment]
I make required changes, but I'm new to github lifecycle so I'm not sure about the results of pull request changes. 
Tell me if any further change or action is required.

**Comment by igitur** [comment]
@roby70 Don't worry, it's a steep learning curve. I'm also a bit stricter than some other repos, but it's necessary for such a large library.

File: ClosedXML.sln [hunk 3]
```diff
@@ -1,7 +1,7 @@
 ﻿
 Microsoft Visual Studio Solution File, Format Version 12.00
-# Visual Studio 14
-VisualStudioVersion = 14.0.25420.1
+# Visual Studio 15
```
Comment (igitur) [inline ClosedXML.sln]: I think all the changes in this file can be reverted.

File: ClosedXML_Tests/ClosedXML_Tests.csproj [hunk 4]
```diff
@@ -308,6 +308,9 @@
   <ItemGroup>
     <EmbeddedResource Include="Resource\Images\ImageHandling.png" />
   </ItemGroup>
+  <ItemGroup>
```
Comment (igitur) [inline ClosedXML_Tests/ClosedXML_Tests.csproj]: This part can be reverted?

File: ClosedXML_Tests/Excel/CalcEngine/MathTrigTests.cs [hunk 5]
```diff
@@ -1254,6 +1302,7 @@ public void SumIfs_ReturnsCorrectValues_ReferenceExample2FromMicrosoft(int expec
             }
         }
 
+
```
Comment (igitur) [inline ClosedXML_Tests/Excel/CalcEngine/MathTrigTests.cs]: Really trivial, but please remove this new line that was added.

**Comment by igitur** [comment]
There's no rush. Let me know if you need help.

**Comment by roby70** [comment]
Now it seems ok.

**Comment by igitur** [comment]
Thanks for the contribution.

Sidenote: I see that one of the other repositories you forked, is QLNet. I'm one of the maintainers on that lib too. Are you a quant?

**Comment by roby70** [comment]
I forked QLNet years ago to evaluate it as an alternative to original QuantLib, but then the project lost priority. 
Main use of QL was to price bond. I can't exclude in the future to restart with QL.

**Pull request merged**